Add mTLS hello world example in C++

Ryanfsdf opened the pull request:
Added an mTLS hello world example in C++. I put the example in the official grpc examples directory, I'm not sure if this is the right place to put it.

Much of this code was taken from the current non-mtls hello world example. There were some things that were in the original example which I've kept in this example, though I don't think there's a clear use for them. ex. 
```
grpc::EnableDefaultHealthCheckService(true);
grpc::reflection::InitProtoReflectionServerBuilderPlugin();
```
I've also kept a lot of the original comments and error messages, I'm not sure if it's worth rewording for greater clarity.

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_client.cc:
```diff
@@ -0,0 +1,119 @@
+using helloworld::Greeter;
+
+ABSL_FLAG(std::string, server_address, "localhost:50051",
+  "address of the server");
```
Can you make the help messages into full sentences? E.g. `The address of the gRPC greeter server.`

matthewstevenson88 commented on examples/BUILD:
```diff
@@ -107,6 +119,19 @@ cc_binary(
+    deps = [
+        ":helloworld_cc_grpc",
+        "//:grpc++",
+        "//:grpc++_reflection",
```
Is this dependency needed? I think this might get caught by ClangTidy.

Ryanfsdf replied:
Yes it's needed, since the server uses `grpc::reflection::InitProtoReflectionServerBuilderPlugin();`.

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_client.cc:
```diff
@@ -0,0 +1,119 @@
+
+int main(int argc, char** argv) {
+  absl::ParseCommandLine(argc, argv);
+  std::string serverAddr = absl::GetFlag(FLAGS_server_address);
```
Could we consolidate reading the flags? For example, have

`sslOpts.pem_root_certs = readFile(absl::GetFlag(...))`

Ryanfsdf replied:
I agree with consolidating `readFile` with the assignment. ie
```
sslOpts.pem_root_certs = readFile(rootCerts);
```
but I don't think we should also consolidate `GetFlag` inside `readFile` as well. I think it would be cleaner and easier to read if we do all the flag reading right at the beginning, otherwise we'll end up with numerous places in the file where flag reading occurs. WDYT?

matthewstevenson88 replied:
I'm not sure I understand your point. Isn't each variable used only once? So if we omitted these variables and just read the flags where they are needed, then we would only be reading each flag once?

Ryanfsdf replied:
Oh I meant that out of the 4 flags, 3 of them would be read at the beginning and 1 of them (server_address) later on in the code. 
```
  sslOpts.pem_root_certs = readFile(GetFlag(...));
  sslOpts.pem_private_key = readFile(GetFlag(...));
  sslOpts.pem_cert_chain = readFile(GetFlag(...));
...
...
...
  example = GetFlag(...)
```
I'll change it to your suggestion, I don't have a strong preference either way!

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_client.cc:
```diff
@@ -0,0 +1,119 @@
+    HelloRequest request;
+    request.set_name(user);
+
+    // Container for the data we expect from the server.
```
[Optional] Feel free to take out the comments in `SayHello`, since this will be used for testing and not instructional purposes. Up to you though.

Ryanfsdf replied:
Done

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_client.cc:
```diff
@@ -0,0 +1,119 @@
+
+int main(int argc, char** argv) {
+  absl::ParseCommandLine(argc, argv);
+  std::string serverAddr = absl::GetFlag(FLAGS_server_address);
```
Please write all variables names not in camel case and using full when words possible. For example, `serverAddr` would become `server_address`.

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_client.cc:
```diff
@@ -0,0 +1,119 @@
+  sslOpts.pem_root_certs = rootCert;
+  sslOpts.pem_private_key = clientKey;
+  sslOpts.pem_cert_chain = clientCert;
+  auto creds = grpc::SslCredentials(sslOpts);
```
Can this line be moved to the channel creation? I.e. have `grpc::CreateChannel(serverAddr, grpc::SslCredentials(...))`.

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_client.cc:
```diff
@@ -0,0 +1,119 @@
+  std::string reply = greeter.SayHello(user);
+  std::cout << "Greeter received: " << reply << std::endl;
+
+  return 0;
```
Can you check that the message you receive is what you expect and if it is return 0; else return 1?

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_server.cc:
```diff
@@ -0,0 +1,117 @@
+// Logic and data behind the server's behavior.
+class GreeterServiceImpl final : public Greeter::Service {
+  Status SayHello(ServerContext* context, const HelloRequest* request,
+                  HelloReply* reply) override {
```
Please add checks to ensure `|reply|` and |request|` are not nullptr.

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_server.cc:
```diff
@@ -0,0 +1,117 @@
+int main(int argc, char** argv) {
+  absl::ParseCommandLine(argc, argv);
+  RunServer();
+
```
Please remove the unused line.

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_server.cc:
```diff
@@ -0,0 +1,117 @@
+  }
+};
+
+std::string readFile(const std::string filePath) {
```
Please make this function `static`.

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_client.cc:
```diff
@@ -0,0 +1,119 @@
+  std::unique_ptr<Greeter::Stub> stub_;
+};
+
+std::string readFile(const std::string& filePath) {
```
Please make this function `static`.

matthewstevenson88 commented:
Thanks for the PR Ryan! It looks great! I've done a first pass and left some comments. Let me know when it's ready for another pass!

Can you run the following 2 commands and make any changes that they point out?

`./tools/distrib/clang_format_code.sh`

and

`./tools/distrib/clang_tidy_code.sh`

Also, did you double check that your C++ client/server can successfully talk to your Go client/server? It looks like there should be no problem but just wanted to confirm.

Ryanfsdf commented:
Thanks for the review! I addressed all the review comments and ran the clang tidy/format scripts.

I had to make some changes to the Dockerfiles to make those scripts work, I'm not sure if it's an issue with my machine specifically. I also noticed that the script itself doesn't automatically run on the `examples` directory, so I had to add that to the script. After running those on the entire `examples/cpp` directory, it seems like there's a lot of linting errors and suggestions on the other example files which I haven't touched. I'm not sure if it's worth fixing in general.

As for the mtls client and server I added, I've fixed all the suggestions from the clang scripts.

Also, I tried cross communicating the client/server with the Go versions but they weren't able to. This is because I didn't set the TLS version to 1.3 in C++. I couldn't find any config/option to do that in C++, do you know where it could be?

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_server.cc:
```diff
@@ -0,0 +1,109 @@
+}
+
+void RunServer() {
+  std::string port = absl::GetFlag(FLAGS_port);
```
Please make this into the full server address.

Ryanfsdf replied:
Shouldn't the server only have it's port be configurable? Since the server will always have it's address as localhost + the port number.

matthewstevenson88 replied:
Since this is what we also do for Go, let's leave this as is.

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_server.cc:
```diff
@@ -0,0 +1,109 @@
+          readFile(server_key_path), readFile(server_cert_path)});
+  sslOpts.pem_root_certs = readFile(root_cert_path);
+
+  grpc::EnableDefaultHealthCheckService(true);
```
I don't think these 2 lines are necessary (for example, we don't need health checking since this server will only be running in a test framework.)

Ryanfsdf replied:
Ack, I removed them

matthewstevenson88 commented:
Thanks for the changes Ryan!

RE forcing gRPC to use TLS 1.3: at present, this is not configurable (this is something I'm working on at the moment). However, for our testing, we can have you make the changes to the gRPC stack needed.

For these changes, look at my [PR #22956](https://github.com/grpc/grpc/pull/22956) and make the changes in the `src/core/tsi/ssl_transport_security.cc` file. In those changes, also change the min TLS version from TLS 1.2 to TLS 1.3. If this is unclear, please ping me and I can explain in more detail. :)

Ryanfsdf commented:
Verified that cross communication with the Go Client/Server works as expected using TLS1.3. PTAL!

matthewstevenson88 commented on examples/cpp/helloworld/mtls_greeter_client.cc:
```diff
@@ -0,0 +1,99 @@
+    if (status.ok()) {
+      return reply.message();
+    } else {
+      std::cout << status.error_code() << ": " << status.error_message()
```
Can you pass this to `std::cerr` rather than `cout`?

matthewstevenson88 commented:
Thanks Ryan! Please add back the changes to `ssl_transport_security.cc`. We want to keep those and not rely on syncing with the main gRPC repo when my PR is merged (which might be a long way off).

Ryanfsdf commented:
Thanks for the review! I added back `ssl_transport_security.cc` and changed `std::cout` to `std::cerr`.
